### PR TITLE
Feature/task error details

### PR DESCRIPTION
**Pull request by andre-merzky**
This PR makes the reporting of task errors outside of the logging mechanism more consistent: exceptions and exception details are exposed as task attributes, the `task.stdout` field will not be overloaded with this information anymore but is now fully reserved for actual task output (starting from the launch script level).

Note that radical-cybertools/radical.saga/pull/857 makes the resulting information much more useful when the exceptions originate on the SAGA level.

### CODE REVIEW

**Comment by andre-merzky** [comment]
Dammit - I responded to comments a couple of days ago (46d3f9803), but somehow my comments on this PR went AWOL.  Sorry for that - but I think I addressed all of it by now.

File: src/radical/pilot/agent/scheduler/base.py [hunk 1]
```diff
@@ -709,7 +713,7 @@ def _schedule_waitpool(self):
       #                                           len(unscheduled), len(failed))
 
         for task, error in failed:
-            task['stderr']       = error
+            task['exception']    = 'RuntimeError(%s)' % error
```
Comment (mtitov) [inline src/radical/pilot/agent/scheduler/base.py:716]: quotation missed: `task['exception'] = 'RuntimeError("%s")' % error`

File: src/radical/pilot/agent/executing/popen.py [hunk 2]
```diff
@@ -102,16 +102,10 @@ def work(self, tasks):
...
-                # by the task), but it seems the most intuitive way to
-                # communicate that error to the application/user.
+            except Exception as e:
                 self._log.exception("error running Task")
-                if task['stderr'] is None:
-                    task['stderr'] = ''
-                task['stderr'] += '\nPilot cannot start task:\n'
-                task['stderr'] += '\n'.join(ru.get_exception_trace())
+                task['exception']        = repr(e)
+                task['exception_detail'] = '\n'.join(ru.get_exception_trace())
```
Comment (mtitov) [inline src/radical/pilot/agent/executing/popen.py:108]: Should the similar update be applied to the Flux executor? ([here](https://github.com/radical-cybertools/radical.pilot/blob/dfb4a2ce38de27c591da43eaaae110cc92ba9b15/src/radical/pilot/agent/executing/flux.py#L264-L267))

File: src/radical/pilot/worker/stager.py [hunk 3]
```diff
@@ -77,8 +77,9 @@ def work(self, sds):
 
         except Exception as e:
             for sd in sds:
-                sd['state'] = rps.FAILED
-                sd['error'] = str(e)
+                sd['exception']        = repr(e)
+                sd['exception_detail'] = '\n'.join(ru.get_exception_trace())
+                sd['state']            = rps.FAILED
```
[the remark this answers is not in the document]
Comment (mtitov) [reply src/radical/pilot/worker/stager.py:82]: the last point here - should we add `self._active_sds[uid]['exception'] = sd['exception']` to PMGRLaunching [here](https://github.com/radical-cybertools/radical.pilot/blob/dfb4a2ce38de27c591da43eaaae110cc92ba9b15/src/radical/pilot/pmgr/launching/base.py#L1108) (same as set for PilotManager)

[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/worker/stager.py:82]: That is made consistent now - thanks.

File: src/radical/pilot/staging_directives.py [hunk 4]
```diff
@@ -67,13 +67,17 @@ def expand_staging_directives(sds: Union[str, Dict[str, Any], List[str]]
...
-            expanded = {'source':   src.strip(),
-                        'target':   tgt.strip(),
-                        'action':   DEFAULT_ACTION,
-                        'flags':    DEFAULT_FLAGS,
-                        'priority': DEFAULT_PRIORITY,
-                        'uid':      ru.generate_id('sd.%(item_counter)06d',
-                                                    ru.ID_CUSTOM, ns='foo')}
+            expanded = {
+                    'uid'             : ru.generate_id('sd.%(item_counter)06d',
+                                                       ru.ID_CUSTOM, ns='foo'),
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/staging_directives.py]: That has been fixed - thanks.

File: src/radical/pilot/raptor/worker_mpi.py [hunk 5]
```diff
@@ -472,20 +474,23 @@ def run(self):
...
+                        task['exception_detail'] = exc[1]
 
                     except Exception as e:
-                        task['error']        = repr(e)
-                        task['stdout']       = ''
-                        task['stderr']       = str(e)
-                        task['exit_code']    = -1
-                        task['return_value'] = None
-                        task['exception']    = [e.__class__.__name__, str(e)]
+                        task['error']            = repr(e)
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/raptor/worker_mpi.py]: That has been fixed also.

File: src/radical/pilot/agent/scheduler/base.py [hunk 6]
```diff
@@ -853,10 +858,11 @@ def _schedule_incoming(self):
 
             except Exception as e:
 
-                task['stderr']       = str(e)
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/agent/scheduler/base.py:856]: Removed it.

File: src/radical/pilot/agent/scheduler/base.py (hunk 1 above)
Comment (andre-merzky) [reply src/radical/pilot/agent/scheduler/base.py:716]: Nice catch!  Fixed now.

File: src/radical/pilot/agent/executing/popen.py (hunk 2 above)
Comment (andre-merzky) [reply src/radical/pilot/agent/executing/popen.py:108]: Fair - that's done now.

**Comment by andre-merzky** [comment]
Thanks Mikhail, those were useful and I addressed them now.

**Review by mtitov — APPROVED** [review]
LGTM!

**Pull request merged**